File: sdk/src/client/bindings/python/examples/11_build_output.py
from iota_client import *
import json

# Create an IotaClient instance
client = IotaClient(nodes = ['https://api.testnet.shimmer.network'])

hex_address = client.bech32_to_hex('rms1qpllaj0pyveqfkwxmnngz2c488hfdtmfrj3wfkgxtk4gtyrax0jaxzt70zy')

address_unlock_condition = AddressUnlockCondition(
    Ed25519Address(hex_address)
)

# Build most basic output with amound and a single address unlock condition
basic_output = client.build_basic_output(
    unlock_conditions=[
        address_unlock_condition
    ],
    amount=1000000,
)
print(json.dumps(basic_output, indent=4))

# Output with metadata feature block
basic_output = client.build_basic_output(
    unlock_conditions=[
        address_unlock_condition,
    ],
    features=[
        MetadataFeature('0x'+'Hello, World!'.encode('utf-8').hex())
    ],
    amount=1000000,
)
print(json.dumps(basic_output, indent=4))

# Output with storage deposit return
basic_output = client.build_basic_output(
    unlock_conditions=[
        address_unlock_condition,
        StorageDepositReturnUnlockCondition(
            return_address=Ed25519Address(hex_address),
            amount=1000000
        )
    ],
    amount=1000000,
)
print(json.dumps(basic_output, indent=4))

# Output with expiration
basic_output = client.build_basic_output(
    unlock_conditions=[
        address_unlock_condition,
        ExpirationUnlockCondition(
            return_address=Ed25519Address(hex_address),
            unix_time=1
        )
    ],
    amount=1000000,
)
print(json.dumps(basic_output, indent=4))

# Output with timelock
basic_output = client.build_basic_output(
    unlock_conditions=[
        address_unlock_condition,
        TimelockUnlockCondition(
            unix_time=1
        )
    ],
    amount=1000000,
)
print(json.dumps(basic_output, indent=4))

File: bindings/nodejs/examples/client/11-build-output.ts
// Copyright 2021-2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {
    Client,
    initLogger,
    Utils,
    UnlockCondition,
    AddressUnlockCondition,
    MetadataFeature,
    SenderFeature,
    TagFeature,
    StorageDepositReturnUnlockCondition,
    ExpirationUnlockCondition,
    TimelockUnlockCondition,
    utf8ToHex,
} from '@iota/sdk';
require('dotenv').config({ path: '.env' });

// Run with command:
// yarn run-example ./client/11-build-output.ts

// In this example we will build a basic output with various options.
async function run() {
    initLogger();
    for (const envVar of ['NODE_URL']) {
        if (!(envVar in process.env)) {
            throw new Error(`.env ${envVar} is undefined, see .env.example`);
        }
    }

    const client = await Client.create({
        // Insert your node URL in the .env.
        nodes: [process.env.NODE_URL as string],
    });

    try {
        const ed25519Address = Utils.parseBech32Address(
            'rms1qpllaj0pyveqfkwxmnngz2c488hfdtmfrj3wfkgxtk4gtyrax0jaxzt70zy',
        );

        const addressUnlockCondition: UnlockCondition =
            new AddressUnlockCondition(ed25519Address);

        // Build most basic output with amount and a single address unlock condition
        const basicOutput = await client.buildBasicOutput({
            amount: BigInt(1000000),
            unlockConditions: [addressUnlockCondition],
        });

        console.log(JSON.stringify(basicOutput, null, 2));

        // Output with metadata feature block
        const basicOutputWithMetadata = await client.buildBasicOutput({
            amount: BigInt(1000000),
            unlockConditions: [addressUnlockCondition],
            features: [
                new MetadataFeature({ data: utf8ToHex('Hello World!') }),
            ],
        });

        console.log(JSON.stringify(basicOutputWithMetadata, null, 2));

        // Output with storage deposit return
        const basicOutputWithStorageReturn = await client.buildBasicOutput({
            amount: BigInt(1000000),
            unlockConditions: [
                addressUnlockCondition,
                new StorageDepositReturnUnlockCondition(
                    ed25519Address,
                    '1000000',
                ),
            ],
        });

        console.log(JSON.stringify(basicOutputWithStorageReturn, null, 2));

        // Output with expiration
        const basicOutputWithExpiration = await client.buildBasicOutput({
            amount: BigInt(1000000),
            unlockConditions: [
                addressUnlockCondition,
                new ExpirationUnlockCondition(ed25519Address, 1),
            ],
        });

        console.log(JSON.stringify(basicOutputWithExpiration, null, 2));

        // Output with timelock
        const basicOutputWithTimelock = await client.buildBasicOutput({
            amount: BigInt(1000000),
            unlockConditions: [
                addressUnlockCondition,
                new TimelockUnlockCondition(1),
            ],
        });

        console.log(JSON.stringify(basicOutputWithTimelock, null, 2));

        // Output with tag feature
        const basicOutputWithTag = await client.buildBasicOutput({
            amount: BigInt(1000000),
            unlockConditions: [addressUnlockCondition],
            features: [new TagFeature(utf8ToHex('Hello, World!'))],
        });

        console.log(JSON.stringify(basicOutputWithTag, null, 2));

        // Output with sender feature
        const basicOutputWithSender = await client.buildBasicOutput({
            amount: BigInt(1000000),
            unlockConditions: [addressUnlockCondition],
            features: [new SenderFeature(ed25519Address)],
        });

        console.log(JSON.stringify(basicOutputWithSender, null, 2));
    } catch (error) {
        console.error('Error: ', error);
    }
}

void run().then(() => process.exit());
